🏃E2E Validation test

wfernandes opened the pull request:
Signed-off-by: Warren Fernandes <wfernandes@pivotal.io>

<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, minor or feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🏃 (:running:, other) -->

**What this PR does / why we need it**:
This PR uses the e2e test framework for e2e tests within cluster-api. I'm hoping this PR accomplishes the following:
- Remove redundant/dead code from cluster-api
- Provide feedback to improve the e2e test framework
- Add e2e tests for test cases that makes sense within cluster-api such as proper webhook setup.

It's currently WIP as I'm having some client/scheme issues. 
To run the tests `make test-e2e`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1830

wfernandes commented on test/e2e/e2e_suite_test.go:
```diff
@@ -55,87 +57,157 @@ var (
+	// TODO: Can the generator handle the install and waiting for its
+	// components. Instead of calling it component generator, we can call it
+	// component installer.
+	framework.WaitForAPIServiceAvailable(ctx, kindCluster, "v1beta1.webhook.cert-manager.io")
```
e2e test framework feedback:
As part of component installation we should have a "post-check" to verify components have been installed correctly. Instead of having something like "component generator" we could just create "component installers" that would encapsulate this behavior. That way we can pass the "installers" to `framework.InstallComponents(ctx, kindCluster, certManInstaller, capiInstaller, capdInstaller)`
Each installer would encapsulate post-check behavior like `WaitForAPIServiceAvailable` for `certManInstaller` and `WaitForPodsReadyInNamespace` for `capiInstaller`.

Discuss with @chuckha

wfernandes commented:
Currently when I run `make test-e2e` I get the following error:
```
...
Cluster API Validation
/Users/wfernandes/workspace/cluster-api/test/e2e/validation_test.go:35
  verifies validation is setup correctly and works [It]
  /Users/wfernandes/workspace/cluster-api/test/e2e/validation_test.go:37

  Unexpected error:
      <*meta.NoKindMatchError | 0xc0000a5e40>: {
          GroupKind: {
              Group: "cluster.x-k8s.io",
              Kind: "Machine",
          },
          SearchedVersions: ["v1alpha3"],
      }
      no matches for kind "Machine" in version "cluster.x-k8s.io/v1alpha3"
  occurred

  /Users/wfernandes/workspace/cluster-api/test/e2e/validation_test.go:76

  Full Stack Trace
...
```
I've checked that the scheme associated with the client does in fact have the kind so **I'm honestly not sure what is going on**. 
See https://github.com/wfernandes/cluster-api/blob/86d177d460e118fa35dfc0b4e46fb4cad9db54ea/test/e2e/e2e_suite_test.go#L96-L98

Any insights are appreciated :)

ncdc commented on test/framework/management/kind/mgmt.go:
```diff
@@ -179,6 +179,12 @@ func getKubeconfigPath(ctx context.Context, name string) (string, error) {
+		fmt.Printf("%q : %q\n", k, v)
+	}
+	fmt.Println("===============")
 	cl, err := client.New(restConfig, client.Options{Scheme: c.Scheme})
```
You need to create `api.NewDynamicRESTMapper` and pass it in to the options as the `Mapper`

wfernandes replied:
That worked! Thanks so much for the tip! I was stuck on this problem for a few days now.

wfernandes commented on Makefile:
```diff
@@ -92,7 +92,9 @@ test-integration: ## Run integration tests
 .PHONY: test-e2e
 test-e2e: ## Run e2e tests
```
Should we add this target to `scripts/ci-test.sh`?

detiber replied:
How long is the runtime? If it's significant we might want to adopt it as a post-submit job rather than a pre-submit job.

wfernandes replied:
Currently, when I run `time make test-e2e` I get:
> real    2m37.973s
user    0m12.329s
sys     0m8.935s

As more e2e tests are added in, it could take longer. So post-submit 🤷‍♂ ? I really wish I knew more about the CI system because I'm not sure what the best option is.

ncdc replied:
How stable is the test? Does it flake at all?

wfernandes replied:
I ran the test with `-count=5` on my machine and it passed without any flakes. But I know from experience that it could be different in a CI environment with constrained resources.

Currently, I've added specific check conditions to ensure that the components are up and running before the test actually runs so I'm confident that there is a low chance for flakes.

If there is concern that there will be additional burden in debugging CI test failures I'm more than happy to share that load. I remember it being brought up during a CAPI meeting some time ago.

detiber replied:
I think with the running time we are talking about doing it during pre-submit should be fine. We can always break it out later if needed.

wfernandes replied:
I created this issue to track this work. https://github.com/kubernetes-sigs/cluster-api/issues/1884

chuckha commented on test/infrastructure/docker/config/default/manager_image_patch.yaml:
```diff
@@ -8,5 +8,5 @@ spec:
       containers:
       # Change the value of image field below to your controller image URL
-      - image: gcr.io/k8s-staging-capi-docker/capd-manager:dev
+      - image: gcr.io/k8s-staging-capi-docker/capd-manager-amd64:dev
```
why is this changing?

wfernandes replied:
Hmm...I didn't change that manually. I believe that this change occurred when I ran `make -C test/infrastructure/docker test-e2e`.
It doesn't need to be in this PR though. I'll take it out.

wfernandes commented:
Work done in this PR: 

1. In `test/e2e/e2e_suite_test.go`
    - Used the e2e test framework to setup a kind cluster and install cert-manager and capi components. CAPI components are installed via `out/cluster-api-components.yaml`
    - Deleted all the redundant helper functions that accomplished the same thing that the framework did.
    - Added `framework.WaitForPodsReadyInNamespace` to ensure that CAPI components are running before moving on to the tests. This increases reliability of the tests. This is added in `test/framework/convenience.go`.
2. In `test/e2e/validation_test.go`
    - Added the two tests that ensures validation webhooks are setup correctly in the environment.
3. In `test/framework/management/kind/mgmt.go`
    - The changes here were to avoid this error (https://github.com/kubernetes-sigs/cluster-api/pull/1859#issuecomment-563368041) when we tried to `kindClient.Create(ctx, machine)` in the validation test. @ncdc suggested the tip (https://github.com/kubernetes-sigs/cluster-api/pull/1859#discussion_r355611841)
4. In the `Makefile`
    - Added appropriate targets that build docker images and generate the manifests in order to deploy CAPI in the e2e test kind cluster.

chuckha commented:
I need some more time with the changes to cluster-api/e2e but i would like the changes to the framework. i lifted the WaitForPod function for my own PR but don't want to steal authorship. Would you be able to pull changes to the framework out and open a PR that we can merge first?

wfernandes commented:
@chuckha Sure. I'm assuming you are referring to the changes in `test/framework`. I'll get on that asap.

chuckha commented on test/e2e/e2e_suite_test.go:
```diff
@@ -24,118 +24,98 @@ import (
 var (
-	managerImage    = flag.String("managerImage", "", "Docker image to load into the kind cluster for testing")
-	capiComponents  = flag.String("capiComponents", "", "URL to CAPI components to load")
-	kustomizeBinary = flag.String("kustomizeBinary", "kustomize", "path to the kustomize binary")
```
Have you seen how the e2es are running on prow? Will we need to provide these features to the framework? namely being able to override various paths/binaries?

wfernandes replied:
Which e2es are these? For the CAPI e2e tests we should just need to run `make test-e2e`. The manager image is being passed in via an env var. 
If we need access to other things for prow/CI, we can add those in as part of #1884 
WDYT?

chuckha replied:
Oh got it, i thought these were already running on prow as jobs

chuckha commented:
/approve
/assign @detiber 

I believe @detiber did most of the work on this e2e code? Assigning for lgtm

detiber commented on test/e2e/e2e_suite_test.go:
```diff
@@ -24,118 +24,98 @@ import (
 
 const (
 	capiNamespace      = "capi-system"
+	capiTestNamespace  = "default"
```
Should we use a randomized name here instead of the default namespace? I worry this might make parallelizing or running the tests against an existing environment more difficult.

wfernandes replied:
@detiber Were you thinking of a `namespace` per `It` test block or just for this suite of tests?

detiber replied:
@wfernandes definitely not per `It`, but maybe a bit more granular than per-suite as well.

detiber commented on test/e2e/validation_test.go:
```diff
@@ -0,0 +1,96 @@
+
+		err := kindClient.Create(ctx, machine)
+		// TODO: Should we assert the error contents to ensure it is a
+		// validation error?
```
+1 to verifying that it is a validation error received.

detiber commented on test/e2e/validation_test.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+var _ = Describe("Cluster API Validation", func() {
+	It("verifies validation is setup correctly and works", func() {
```
Should this be updated to state that it is attempting validation of `Machine` specifically here? I suspect we should also test validation of the other types as well, otherwise we are only verifying that the `Machine` validation webhooks are functional.

wfernandes replied:
Yup. Thanks for pointing this out. I was hoping to add just this test for now and then adding future validation tests for other types.

detiber replied:
sgtm

wfernandes commented on test/e2e/validation_test.go:
```diff
@@ -19,78 +19,94 @@ limitations under the License.
+		framework.EnsureNamespace(context.Background(), kindClient, capiTestNamespace)
+	})
+
+	Context("for Machine", func() {
```
This should create a namespace for each context block. That way it is granular to specific validation hook types.

wfernandes commented on test/e2e/validation_test.go:
```diff
@@ -19,78 +19,94 @@ limitations under the License.
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(Equal(
+				fmt.Sprintf(`admission webhook "validation.machine.cluster.x-k8s.io" denied the request: Machine.cluster.x-k8s.io "%s" is invalid: spec.bootstrap.data: Required value: expected spec.bootstrap.data or spec.bootstrap.configRef to be populated`, name),
+			))
```
I tried to inspect the `*apierrors.StatusError` to see if there was a way to see if this was a "Validation" failure but there wasn't such a distinction on the error object.

detiber commented:
/lgtm
